### PR TITLE
Add support for HEIC uploading + JPEG conversion

### PR DESCRIPTION
**Pull request by adamsilverstein**
First pass, excluding tests

Trac ticket: https://core.trac.wordpress.org/ticket/53645

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/js/_enqueues/vendor/plupload/handlers.js [hunk 1]
```diff
@@ -613,6 +613,11 @@ jQuery( document ).ready( function( $ ) {
 					wpQueueError( pluploadL10n.noneditable_image );
 					up.removeFile( file );
 					return;
+				} else if ( file.type === 'image/heic' && up.settings.heic_upload_error ) {
```
Comment (noisysocks) [inline src/js/_enqueues/vendor/plupload/handlers.js]: This condition already exists above on line 603.

File: src/wp-admin/includes/media.php [hunk 2]
```diff
@@ -2204,6 +2204,11 @@ function media_upload_form( $errors = null ) {
 		$plupload_init['avif_upload_error'] = true;
 	}
 
+	// Check if HEIC images can be edited.
+	if ( ! wp_image_editor_supports( array( 'mime_type' => 'image/heic' ) ) ) {
+		$plupload_init['heic_upload_error'] = true;
```
Comment (noisysocks) [inline src/wp-admin/includes/media.php]: `heic_upload_error` is already set in `wp_show_heic_upload_error` which is hooked into the `plupload_init` filter.

This seems the more logical place, though, so maybe we can deprecate `wp_show_heic_upload_error`.

File: tests/phpunit/tests/media.php [hunk 3]
```diff
@@ -5290,6 +5290,51 @@ public function test_quality_with_image_conversion_file_sizes() {
 		}
 	}
 
+	/**
+	 * Test AVIF quality filters.
+	 *
+	 * @ticket 61614
+	 */
+	public function test_quality_with_avif_conversion_file_sizes() {
```
Comment (noisysocks) [inline tests/phpunit/tests/media.php]: Is this test relevant to Trac-53645? Might need you to explain to me what this is for 😀

**Comment by noisysocks** [comment]
Thanks for the PR @adamsilverstein. It doesn't seem to work when I test it, though. 

If I drop a .heic into the gallery I see an error, the image cannot be previewed, and the file type is still `image/heic`.

https://github.com/user-attachments/assets/6a11403b-c794-439c-8cb6-59ebf104817f

File: src/wp-admin/includes/media.php (hunk 2 above)
Comment (adamsilverstein) [reply src/wp-admin/includes/media.php]: good point, I was just trying to copy all the places it was added for avif, trying to exclude those having to do with display (we aren't trying to display HEIC, only convert uploads to JPEG).

File: tests/phpunit/tests/media.php (hunk 3 above)
Comment (adamsilverstein) [reply tests/phpunit/tests/media.php]: no this test is for Trac-61614 and snuck in because I didn't clean up completely, I will remove it. Thanks!

File: src/js/_enqueues/vendor/plupload/handlers.js (hunk 1 above)
Comment (adamsilverstein) [reply src/js/_enqueues/vendor/plupload/handlers.js]: good point, I will leave the current behavior and remove this bit

File: src/wp-admin/includes/media.php (hunk 2 above)
Comment (adamsilverstein) [reply src/wp-admin/includes/media.php]: Oh, I see now. Let me check for any plugin use of `wp_show_heic_upload_error` and consider deprecating it. Otherwise, I'll at least add the mime type supports check there.

File: src/js/media/controllers/library.js [hunk 4]
```diff
@@ -196,7 +196,7 @@ Library = wp.media.controller.State.extend(/** @lends wp.media.controller.Librar
 	isImageAttachment: function( attachment ) {
 		// If uploading, we know the filename but not the mime type.
 		if ( attachment.get('uploading') ) {
-			return /\.(jpe?g|png|gif|webp|avif)$/i.test( attachment.get('filename') );
+			return /\.(jpe?g|png|gif|webp|avif|heif)$/i.test( attachment.get('filename') );
```
Comment (swissspidy) [inline src/js/media/controllers/library.js]: The file extension for HEIF images is typically `.heic`, but there's also `.heics`,`.heif` and `.heifs`. The "s" is for sequences.

File: src/wp-includes/media.php [hunk 5]
```diff
@@ -4061,7 +4061,7 @@ function wp_get_image_editor( $path, $args = array() ) {
 	// Check and set the output mime type mapped to the input type.
 	if ( isset( $args['mime_type'] ) ) {
 		/** This filter is documented in wp-includes/class-wp-image-editor.php */
-		$output_format = apply_filters( 'image_editor_output_format', array(), $path, $args['mime_type'] );
+		$output_format = apply_filters( 'image_editor_output_format', array( 'image/heif' => 'image/jpeg' ), $path, $args['mime_type'] );
```
Comment (swissspidy) [inline src/wp-includes/media.php]: Isn't it `image/heic`?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php [hunk 6]
```diff
@@ -531,7 +531,7 @@ public function edit_media_item( $request ) {
 			);
 		}
 
-		$supported_types = array( 'image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/avif' );
+		$supported_types = array( 'image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/avif', 'image/heif' );
```
Comment (swissspidy) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php]: Isn't it `image/heic`? Both seem to be used interchangeably here and there.

File: src/wp-includes/media.php (hunk 5 above)
Comment (spacedmonkey) [inline src/wp-includes/media.php]:
```suggestion
		$output_format = apply_filters( 'image_editor_output_format', array( 'image/heic' => 'image/jpeg' ), $path, $args['mime_type'] );
```

File: src/wp-includes/functions.php [hunk 7]
```diff
@@ -2707,7 +2707,7 @@ function wp_unique_filename( $dir, $filename, $unique_filename_callback = null )
 		 */
 		if ( $is_image ) {
 			/** This filter is documented in wp-includes/class-wp-image-editor.php */
-			$output_formats = apply_filters( 'image_editor_output_format', array(), $_dir . $filename, $mime_type );
+			$output_formats = apply_filters( 'image_editor_output_format', array( 'image/heif' => 'image/jpeg' ), $_dir . $filename, $mime_type );
```
Comment (spacedmonkey) [inline src/wp-includes/functions.php]:
```suggestion
			$output_formats = apply_filters( 'image_editor_output_format', array( 'image/heic' => 'image/jpeg' ), $_dir . $filename, $mime_type );
```

File: src/wp-includes/class-wp-image-editor.php [hunk 8]
```diff
@@ -375,17 +374,18 @@ protected function get_output_format( $filename = null, $mime_type = null ) {
...
-		 *     destination mime type. Default empty array.
+		 *     destination mime type. Default maps uploaded HEIC images to JPEG output.
 		 *
 		 *     @type string ...$0 The new mime type.
 		 * }
 		 * @param string $filename  Path to the image.
 		 * @param string $mime_type The source image mime type.
 		 */
-		$output_format = apply_filters( 'image_editor_output_format', array(), $filename, $mime_type );
+		$output_format = apply_filters( 'image_editor_output_format', array( 'image/heif' => 'image/jpeg' ), $filename, $mime_type );
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-image-editor.php]:
```suggestion
		$output_format = apply_filters( 'image_editor_output_format', array( 'image/heic' => 'image/jpeg' ), $filename, $mime_type );
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php (hunk 6 above)
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php]:
```suggestion
		$supported_types = array( 'image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/avif', 'image/heic' );
```

File: src/js/media/controllers/library.js (hunk 4 above)
Comment (adamsilverstein) [reply src/js/media/controllers/library.js]: ah right, for this PR I'm trying to add just support for `.heic` files which I though is what iphones use as their default output format (looking now). Do you think there is value is supporting all of the possible extensions?

File: src/wp-admin/includes/media.php (hunk 2 above)
Comment (adamsilverstein) [reply src/wp-admin/includes/media.php]: nothing that would break plugins, however I see that when this filter was added, @azaozz commented that is specifically added via filters so that it can be removed. for that reason, I'm going to move this code to the filter and remove the use here.

Comment (adamsilverstein) [reply src/wp-admin/includes/media.php]: Fixed in 683a076401d221a65ee2657e92897b00086314cc

File: src/wp-includes/media.php (hunk 5 above)
Comment (adamsilverstein) [reply src/wp-includes/media.php]: right, fixing!

File: src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php (hunk 6 above)
Comment (adamsilverstein) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php]: right, fixing now...

File: src/js/media/controllers/library.js (hunk 4 above)
Comment (swissspidy) [reply src/js/media/controllers/library.js]: I don't think so, haven't seen those in the wild yet.

**Comment by adamsilverstein** [comment]
> Thanks for the PR @adamsilverstein. It doesn't seem to work when I test it, though.

@noisysocks some parts were missing, I had a little trouble debugging, but finally got it working. Can you give it another test?

**Comment by adamsilverstein** [comment]
Here is a screencast of uploading an HEIC image via drag and drop onto the editor:

https://github.com/user-attachments/assets/600766e8-bb07-4fa8-ad88-3dfad79139e1

on the right side you can see all of the files being generated in jpeg format. Note the preview functionality still isn't working correctly.

Also, using the "upload" selector in Gutenberg fails because HEIC is still not a recognized type there:

image

This probably needs fixing in Gutenberg.

**Comment by noisysocks** [comment]
>  Note the preview functionality still isn't working correctly.

Yep we'll need to fix those two issues in the Gutenberg repo.

> Also, using the "upload" selector in Gutenberg fails because HEIC is still not a recognized type there:

This seems to work for me. Might be browser dependent. We can investigate on the Gutenberg side.

**Comment by noisysocks** [comment]
I'm seeing a different error about the image size now when I test with a picture I took the other day.

https://github.com/user-attachments/assets/1643092f-4c6c-40b4-b8d9-a8469cbe68ee

Error log:

```
2024/07/17 05:13:36 [warn] 2197#2197: *2317 a client request body is buffered to a temporary file /var/cache/nginx/client_temp/0000000026, client: 192.168.228.6, server: , request: "POST /wp-admin/async-upload.php HTTP/1.1", host: "wordpress-develop.ddev.site", referrer: "https://wordpress-develop.ddev.site/wp-admin/upload.php"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "NOTICE: PHP message: PHP Fatal error:  Uncaught Error: Undefined constant "IMAGETYPE_HEIC" in /var/www/html/build/wp-includes/media.php:5607"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "Stack trace:"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "#0 /var/www/html/build/wp-admin/includes/image.php(803): wp_getimagesize('/var/www/html/b...')"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "#1 /var/www/html/build/wp-admin/includes/media.php(392): wp_read_image_metadata('/var/www/html/b...')"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "#2 /var/www/html/build/wp-admin/includes/ajax-actions.php(2632): media_handle_upload('async-upload', NULL, Array)"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "#3 /var/www/html/build/wp-admin/async-upload.php(33): wp_ajax_upload_attachment()"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "#4 {main}"
[17-Jul-2024 05:13:36] WARNING: [pool www] child 10848 said into stderr: "  thrown in /var/www/html/build/wp-includes/media.php on line 5607"
2024/07/17 05:13:36 [error] 2197#2197: *2317 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Undefined constant "IMAGETYPE_HEIC" in /var/www/html/build/wp-includes/media.php:5607
Stack trace:
#0 /var/www/html/build/wp-admin/includes/image.php(803): wp_getimagesize('/var/www/html/b...')
#1 /var/www/html/build/wp-admin/includes/media.php(392): wp_read_image_metadata('/var/www/html/b...')
#2 /var/www/html/build/wp-admin/includes/ajax-actions.php(2632): media_handle_upload('async-upload', NULL, Array)
#3 /var/www/html/build/wp-admin/async-upload.php(33): wp_ajax_upload_attachment()
#4 {main}
  thrown in /var/www/html/build/wp-includes/media.php on line 5607" while reading response header from upstream, client: 192.168.228.6, server: , request: "POST /wp-admin/async-upload.php HTTP/1.1", upstream: "fastcgi://unix:/run/php-fpm.sock:", host: "wordpress-develop.ddev.site", referrer: "https://wordpress-develop.ddev.site/wp-admin/upload.php"
```

Looks like `media.php` references `IMAGETYPE_HEIF` but `combat.php` defines `IMAGETYPE_HEIC`.

**Comment by noisysocks** [comment]
Wiki says `.heic`, `.heif`, `image/heic` and `image/heif` are all valid. Should we be liberal in what we accept? In practice I've only ever seen `.heic` and `image/heic`.

https://en.wikipedia.org/wiki/High_Efficiency_Image_File_Format

File: src/wp-includes/media.php [hunk 9]
```diff
@@ -5582,6 +5592,29 @@ function wp_getimagesize( $filename, ?array &$image_info = null ) {
 		}
 	}
 
+	// For PHP versions that don't support HEIC images, extract the size infor using Imagick when available.
```
Comment (noisysocks) [inline src/wp-includes/media.php]:
```suggestion
	// For PHP versions that don't support HEIC images, extract the size info using Imagick when available.
```

**Comment by noisysocks** [comment]
OK, it's working well for me with this patch applied:

```diff
diff --git a/src/wp-includes/compat.php b/src/wp-includes/compat.php
index f7c1a85871..bd4f1b69dc 100644
--- a/src/wp-includes/compat.php
+++ b/src/wp-includes/compat.php
@@ -514,6 +514,6 @@ if ( ! defined( 'IMG_AVIF' ) ) {
 }
 
 // IMAGETYPE_HEIF constant is not yet defined in PHP as of PHP 8.3.
-if ( ! defined( 'IMAGETYPE_HEIF' ) ) {
-	define( 'IMAGETYPE_HEIF', 99 );
+if ( ! defined( 'IMAGETYPE_HEIC' ) ) {
+	define( 'IMAGETYPE_HEIC', 99 );
 }
```

I tested a bunch of different upload flows. Here's how it looks:

https://github.com/user-attachments/assets/f1c3d19b-fd5d-46e4-a2fc-d35945653a9c

I like that this approach means that you can download the original HEIC image via the attachment page. Very nice.

👍 LGTM once that `define()` is fixed.

**Comment by noisysocks** [comment]
Here's a fix for the Image block: https://github.com/WordPress/gutenberg/pull/63643

**Comment by swissspidy** [comment]
Interestingly, in my project I just saw some differences in the browser tests where the browser-determined mime type was once `image/heic` and once `image/heif`. Strangely that was for the same image 🤔 

I don't know how the mime type detection works in PHP, but it might be worth testing this with a few different HEIC images to see if the `image/heif` mime type also needs to be checked (in addition to just `image/heic`).

HEIC images have a different FourCC code depending on the type:

```
'mif1', // .heic / image/heif
'msf1', // .heic / image/heif-sequence
'heic', // .heic / image/heic
'heix', // .heic / image/heic
'hevc', // .heic / image/heic-sequence
'hevx', // .heic / image/heic-sequence
```

**Comment by adamsilverstein** [comment]
> Looks like media.php references IMAGETYPE_HEIF but combat.php defines IMAGETYPE_HEIC.
Ah, right. sticking with HEIC for now, fixed in a65c937c5e6cb92de8ca76ca36d0d20a44c9430e

**Comment by adamsilverstein** [comment]
> Interestingly, in my project I just saw some differences in the browser tests where the browser-determined mime type was once `image/heic` and once `image/heif`. Strangely that was for the same image 🤔
> 
> I don't know how the mime type detection works in PHP, but it might be worth testing this with a few different HEIC images to see if the `image/heif` mime type also needs to be checked (in addition to just `image/heic`).
> 
> HEIC images have a different FourCC code depending on the type:
> 
> ```
> 'mif1', // .heic / image/heif
> 'msf1', // .heic / image/heif-sequence
> 'heic', // .heic / image/heic
> 'heix', // .heic / image/heic
> 'hevc', // .heic / image/heic-sequence
> 'hevx', // .heic / image/heic-sequence
> ```

Currently I'm using [this bit](https://github.com/WordPress/wordpress-develop/pull/7034/files#diff-5526e2d644a04f0c466d6c749542fdb7952021784e7b1583ff45f9d12601ef52R3382) of code which I haven't really verified works beyond a single test image. This will be important to test throughly because PHP won't properly detect the mime type. 

I plan to add tests next and can bundle several test images to make sure they are identified correctly (and hopefully that we can get the sizes correctly with Imagick). I'll be looking for test images that have a compatible license and are relatively small, contributions welcome!

**Comment by adamsilverstein** [comment]
> Here's a fix for the Image block: [WordPress/gutenberg#63643](https://github.com/WordPress/gutenberg/pull/63643)

Amazing, thank you!

**Comment by spacedmonkey** [comment]
This works well in my testing. I really like that you get to the original image from the attachment screen. 

Screenshot 2024-07-17 at 22 38 48

**Comment by spacedmonkey** [comment]
Worth noting that I tried uploading heif from https://filesampleshub.com/format/image/heif. I was unable to upload and I get this error message. 
Screenshot 2024-07-17 at 22 48 03

**Comment by adamsilverstein** [comment]
> This works well in my testing. I really like that you get to the original image from the attachment screen.
> 
> Screenshot 2024-07-17 at 22 38 48

yum!

**Comment by adamsilverstein** [comment]
> Worth noting that I tried uploading heif from [filesampleshub.com/format/image/heif](https://filesampleshub.com/format/image/heif). I was unable to upload and I get this error message. Screenshot 2024-07-17 at 22 48 03

Thanks for the link to the sample files, that is helpful.

I could try to extend full support to .heif files. Do we know how common these are or where/how they are created? I double checked and the files my iphone was creating before I switched it off were .heic files.

**Comment by noisysocks** [comment]
As an experiment I took a `.heic` image from my iPhone and renamed it to `.heif`. Regardless of whether the file is `.heic` or `.heif` it works the same in Preview and other Apple apps. In both cases the MIME type that `file -I`  spits out is `image/heic`.

```
$ file -I playground-1.HEIF
playground-1.HEIF: image/heic; charset=binary
$ file -I playground-2.HEIC
playground-2.HEIC: image/heic; charset=binary
```

I think it makes sense to have WordPress do the same. Support both `.heic` and `.heif` but map them to `image/heic` in both cases.

**Comment by spacedmonkey** [comment]
Other sample images

https://filesampleshub.com/format/image/heic
https://github.com/tigranbs/test-heic-images/tree/master
https://www.learningcontainer.com/sample-heic-files-download/#
https://github.com/dsoprea/heic-exif-samples

File: src/wp-includes/media.php [hunk 10]
```diff
@@ -5582,6 +5592,29 @@ function wp_getimagesize( $filename, ?array &$image_info = null ) {
 		}
 	}
 
+	// For PHP versions that don't support HEIC images, extract the size info using Imagick when available.
```
Comment (noisysocks) [inline src/wp-includes/media.php:5595]:
```suggestion
	// For PHP versions that support HEIC images, extract the size info using Imagick when available.
```

File: src/wp-includes/class-wp-image-editor.php [hunk 11]
```diff
@@ -375,17 +374,18 @@ protected function get_output_format( $filename = null, $mime_type = null ) {
...
-		 *     destination mime type. Default empty array.
+		 *     destination mime type. Default maps uploaded HEIC images to JPEG output.
 		 *
 		 *     @type string ...$0 The new mime type.
 		 * }
 		 * @param string $filename  Path to the image.
 		 * @param string $mime_type The source image mime type.
 		 */
-		$output_format = apply_filters( 'image_editor_output_format', array(), $filename, $mime_type );
+		$output_format = apply_filters( 'image_editor_output_format', array( 'image/heic' => 'image/jpeg' ), $filename, $mime_type );
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-image-editor.php]: For future maintainability, I'd suggest adding a default filter to do the mapping. This will prevent the need to update the default in three location if it changes in the future. It will allow plugin authors to use a simple `remove_filter()` call if they want to maintain HEIC as HEIC.

If you accept this, a `@see` annotation in the docblock would be lovely

**Comment by noisysocks** [comment]
Added some tests @adamsilverstein.

I think coverage is okay, though I struggled to write one for the changes to `wp_generate_attachment_metadata`. Ended up deciding it's not worth the effort since it more or less is already covered by `test_resize_heic`.

Something strange I noticed is that our PHP 8.3 container in GitHub Actions doesn't have the Imagick extension enabled but the other containers do 🤷‍♂️

File: src/wp-includes/class-wp-image-editor.php (hunk 11 above)
Comment (noisysocks) [reply src/wp-includes/class-wp-image-editor.php]: Plugin authors can `unset( $formats['image/heic'] )` using `image_editor_output_format` to maintain HEIC. I think that's a little clearer than unsetting a default filter.

Agree de-duplicating the default array would be nice for maintainability though. Maybe can add a wrapper function a la `wp_get_image_editor_output_format`?

Comment (peterwilsoncc) [reply src/wp-includes/class-wp-image-editor.php]: > Agree de-duplicating the default array would be nice for maintainability though. Maybe can add a wrapper function a la `wp_get_image_editor_output_format`?

Works for me.

Comment (noisysocks) [reply src/wp-includes/class-wp-image-editor.php]: https://github.com/WordPress/wordpress-develop/pull/7034/commits/db78c272f7b06ba273c04afc8194b5b5351fa9de

**Comment by noisysocks** [comment]
This still tests really well for me. Here's how things look now that https://github.com/WordPress/gutenberg/pull/63643 is merged.

https://github.com/user-attachments/assets/6cfcda0c-d04b-445f-9884-677f539aac2a

I'll commit on Monday if there's no objections. Keen to get something in early for 6.7 so that there's plenty of time for feedback and iteration.

**Comment by noisysocks** [comment]
Committed in r58849.
